stream: fix readable behavior for highWaterMark === 0

lundibundi opened the pull request:
Avoid trying to emit 'readable' due to the fact that
state.length is always >= state.highWaterMark if highWaterMark is 0.
Therefore upon .read(0) call (through .on('readable')) stream assumed
that it has enough data to emit 'readable' even though
state.length === 0 instead of issuing _read(). Which led to the TTY
not recognizing that someone is waiting for the input.

Fixes: https://github.com/nodejs/node/issues/20503
Refs: https://github.com/nodejs/node/pull/18372

##### Checklist

- [x] `make -j4 test` (UNIX) passes
- [x] tests and/or benchmarks are included 
      I put one test under sequential as it uses process.stdin to emulate the use case and I'm not sure if that is correct. 
- [ ] documentation is changed or added (not sure what to change)
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

Edit: I've changed the PR title and description according to changes.

mscdex commented:
/cc @nodejs/streams

lundibundi commented:
ping @mcollina.

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/16075/
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1473/
CITGM (master): https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1474/

lundibundi commented:
Well, CI has failed due to my test... I'm not sure how to handle tty stdin on CI, it seems to close the stdin event before the setTimeout fires, can someone help? Should I `fork` a process for this test and only connect stderr to pass the error?
I've also rebased and decreased delay to 1 ms instead of 1000.

BridgeAR commented on test/parallel/test-readable-single-end.js:
```diff
@@ -0,0 +1,16 @@
+r.push(null);
+
+r.once('readable', common.mustCall());
+r.once('end', common.mustCall());
```
When reading the description of the test it seems like these should be `.on` instead of once?

BridgeAR commented on test/sequential/test-readable-tty-keepalive.js:
```diff
@@ -0,0 +1,29 @@
+const assert = require('assert');
+
+// This test ensures that Node.js will not ignore tty 'readable' subscribers
+// when it's the only tty subscriber and te only thing keeping event loop alive
```
Typo in `the`.

lundibundi commented:
@BridgeAR thanks a lot, I've missed that. Should be fixed now.

BridgeAR commented on test/parallel/test-readable-single-end.js:
```diff
@@ -0,0 +1,16 @@
+
+r.push(null);
+
+r.once('readable', common.mustCall());
```
Should this not also just be `on`?

lundibundi replied:
Thanks, idk why did I think it was okay to put once on both of them. :thinking:

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/16213/
CITGM (master): https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1481/
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1482/

mcollina commented:
Seems good, but CITGM is currently not very healthy: https://github.com/nodejs/build/issues/1429.

mafintosh commented:
I'm not I understand why this is needed for TTYs to work? Can someone clarify this for me? The unneeded readables was breaking of code for me, and as I understand this, it restores that behaivor?

lundibundi commented:
@mafintosh After your comment, I went to double check and found a better way of tackling this problem.

But I also discovered the other thing. Your PR is indeed fixing the issue of empty-readable but the thing is, it suppresses the issue by not emitting the event but there are still multiple calls to `emitReadable_` present. I tried to actually resolve the issue but to no avail, unfortunately. Here is the description, maybe someone can figure it out.

The thing is that with the write pattern as in `test/parallel/test-stream-readable-no-unneeded-readable.js` triggers 'readable' event upon push (as it should) but due to the fact that it only arrives on the nextTick that .once('readable') manages to catch old readable event (because we have already read all of the data of that event in the wrapper) that is now empty and results in empty (and multiple) 'readable' events.
If we look at the debug log (I added a few log statements) we can see that actually after last push (before push(null))
```
STREAM 15878: do read
wrapper read called
STREAM 15878: read undefined
STREAM 15878: need readable true
STREAM 15878: length less than watermark true
STREAM 15878: do read
wrapper read called
STREAM 15878: read undefined
STREAM 15878: need readable true
STREAM 15878: length less than watermark true
STREAM 15878: do read
STREAM 15878: on readable 0 true
STREAM 15878: read undefined
STREAM 15878: need readable true
STREAM 15878: length less than watermark true
STREAM 15878: reading or ended false
STREAM 15878: emit readable rStream
STREAM 15878: flow null
STREAM 15878: emit readable wrapperStream
STREAM 15878: flow true
STREAM 15878: read undefined
STREAM 15878: need readable false
STREAM 15878: length less than watermark true
STREAM 15878: reading or ended false
STREAM 15878: emit readable rStream
STREAM 15878: flow null
STREAM 15878: emit readable wrapperStream
STREAM 15878: flow true
STREAM 15878: read undefined
STREAM 15878: need readable false
STREAM 15878: length less than watermark true
STREAM 15878: reading or ended false
STREAM 15878: readableAddChunk null
STREAM 15878: emit readable rStream
readable called
STREAM 15878: read undefined
STREAM 15878: endReadable false
STREAM 15878: flow null
STREAM 15878: endReadableNT false 0
STREAM 15878: readableAddChunk null
STREAM 15878: emit readable wrapperStream
STREAM 15878: flow true
STREAM 15878: read undefined
STREAM 15878: endReadable false
STREAM 15878: endReadableNT false 0
STREAM 15878: do read
STREAM 15878: readableAddChunk <Buffer 62 61 72>
STREAM 15878: emitReadable null
STREAM 15878: readableAddChunk <Buffer 62 61 72>
STREAM 15878: emitReadable true
STREAM 15878: read undefined
STREAM 15878: need readable false
STREAM 15878: length less than watermark true
STREAM 15878: do read
wrapper read called
STREAM 15878: read undefined
STREAM 15878: need readable true
STREAM 15878: length less than watermark true
STREAM 15878: do read
STREAM 15878: on readable 0 true
STREAM 15878: read undefined
STREAM 15878: need readable true
STREAM 15878: length less than watermark true
STREAM 15878: reading or ended false
STREAM 15878: emit readable rStream
STREAM 15878: flow null
STREAM 15878: emit readable wrapperStream
STREAM 15878: flow true
STREAM 15878: read undefined
STREAM 15878: need readable false
STREAM 15878: length less than watermark true
STREAM 15878: reading or ended false
STREAM 15878: emit readable rStream
STREAM 15878: flow null
STREAM 15878: emit readable wrapperStream
STREAM 15878: flow true
STREAM 15878: read undefined
STREAM 15878: need readable false
STREAM 15878: length less than watermark true
STREAM 15878: reading or ended false
STREAM 15878: readableAddChunk null
STREAM 15878: emit readable rStream
readable called
STREAM 15878: read undefined
STREAM 15878: endReadable false
STREAM 15878: flow null
STREAM 15878: endReadableNT false 0
STREAM 15878: readableAddChunk null
STREAM 15878: emit readable wrapperStream
STREAM 15878: flow true
STREAM 15878: read undefined
STREAM 15878: endReadable false
STREAM 15878: endReadableNT false 0
```
there are 2 more 'emit readable rStream' that were suppressed and I think at least the second one shouldn't have happened at all (the first one is from the latest data push).
Though I may be misunderstanding something and this second read/emit is actually needed?

lundibundi commented:
@mcollina @mafintosh @BridgeAR I'd like another review as I changed this PR's idea.

mcollina commented on lib/_stream_readable.js:
```diff
@@ -274,7 +274,9 @@ function readableAddChunk(stream, chunk, encoding, addToFront, skipChunkCheck) {
   // Also, if we have no data yet, we can stand some more bytes.
   // This is to work around cases where hwm=0, such as the repl.
```
We'll need to adjust this comment.

mcollina commented on lib/_stream_readable.js:
```diff
@@ -383,7 +385,8 @@ Readable.prototype.read = function(n) {
   // the 'readable' event and move on.
```
this comment will need an update.

mcollina commented:
It's not really clear what it is changing here. Can you describe it the change to streams a bit more?

lundibundi commented:
@mcollina cleaned up the PR and added relevant test without the TTY. Yeah, this shouldn't change anything, it's a bug fix, bad wording, sorry. I've updated the description.
Also, `test/parallel/test-readable-single-end.js` can be dropped but I see no harm in having more tests.

P.s. Streams sure are tough, thanks for bearing with me.

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/16336/

CITGM (master): https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1497/
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1498/

mcollina commented:
> P.s. Streams sure are tough, thanks for bearing with me.

 Thanks for taking these issue, they are indeed!

mafintosh commented:
Very nice fix in the end

mcollina commented:
Landed in fe47b8b

mcollina commented:
Thanks @lundibundi for tackling this one!
